File: algorithms/568. Maximum Vacation Days/solution.cc
class Solution {
public:
    int maxVacationDays(vector<vector<int>>& flights, vector<vector<int>>& days) {
        int n = days.size(), K = days[0].size();
        vector<int> best(n, INT_MIN), cur(n);
        best[0] = 0;
        for ( int k = 0; k < K; ++k ) {
            fill(cur.begin(), cur.end(), INT_MIN);
            for ( int i = 0; i < n; ++i ) {
                for ( int j = 0; j < n; ++j ) {
                    if ( flights[i][j] || i == j ) {
                        cur[j] = max(cur[j], best[i] + days[j][k]);
                    }
                }
            }
            best = cur;
        }
        return *max_element(best.begin(), best.end());
    }
};

// 57 / 57 test cases passed.
// Status: Accepted
// Runtime: 149 ms

File: algorithms/568. Maximum Vacation Days/solution.py
class Solution(object):
    def maxVacationDays(self, flights, days):
        """
        :type flights: List[List[int]]
        :type days: List[List[int]]
        :rtype: int
        """
        # @awice
        N, K = len(days), len(days[0])
        dp = [float('-inf')] * N
        dp[0] = 0
        for k in xrange(K):
            cur = [float('-inf')] * N
            for i in xrange(N):
                for j in xrange(N):
                    if flights[i][j] or i == j:
                        cur[j] = max(cur[j], dp[i] + days[j][k])
            dp = cur
        return max(dp)

# 57 / 57 test cases passed.
# Status: Accepted
# Runtime: 2549 ms
